onnxruntime/core/graph/graph.cc

```diff
@@ -411,15 +438,20 @@ common::Status NodeArg::UpdateTypeAndShape(const ONNX_NAMESPACE::TypeProto& inpu
     }
 #endif
 
+#if !defined(DISABLE_OPTIONAL_TYPE)
     case TypeProto::kOptionalType: {
-      if ((utils::HasOptionalTensorType(input_type) && !utils::HasOptionalTensorType(current_type)) ||
-          (!utils::HasOptionalTensorType(input_type) && utils::HasOptionalTensorType(current_type))) {
+      bool is_input_type_optional_tensor_type = utils::HasOptionalTensorType(input_type);
+      bool is_current_type_optional_tensor_type = utils::HasOptionalTensorType(current_type);
+
+      // Check for homogeneity within optional type
+      if ((is_input_type_optional_tensor_type && !is_current_type_optional_tensor_type) ||
+          (!is_input_type_optional_tensor_type && is_current_type_optional_tensor_type)) {
```

Comment:
could it be simplified to `is_input_type_optional_tensor_type != is_current_type_optional_tensor_type`?